### ISSUE TITLE
Re-use temporaries in OIR when possible

### ISSUE THREAD

**Issue by jdahm** [GridTools/gt4py#648]
It would be convenient to have an optimization pass on OIR that reduces the number of temporary fields, as we have seen a number of kernels with very high shared memory usage, leading to either (1) not being able to run the kernel or (2) in the best cases very low occupancy.

Reducing the number of temporaries should be possible by using a [live variable analysis](https://en.wikipedia.org/wiki/Live_variable_analysis).

**Comment by FlorianDeconinck** [comment]
Yes `DaCe` is capable of this - and the current strategy is not to do it in OIR.

**Comment by FlorianDeconinck** [comment]
I am debating with myself - this  could be done in OIR and would benefit all `gt:X` backends, as well as the python backends.

Let leave it on - we need to make progress on `dace:X` first.
